allennlp/nn/beam_search.py

```diff
@@ -381,3 +409,178 @@ def _search(
+            of the predicted sequences.
+        beam_size : `int`, optional (default = `10`)
+            The width of the beam used.
+        per_node_beam_size : `int`, optional (default = `beam_size`)
+            The maximum number of candidates to consider per node, at each step in the search.
+            If not given, this just defaults to `beam_size`. Setting this parameter
+            to a number smaller than `beam_size` may give better results, as it can introduce
+            more diversity into the search. See
+            [*Beam Search Strategies for Neural Machine Translation*, Freitag and Al-Onaizan, 2017]
+            (https://arxiv.org/abs/1702.01806).
+        sampler : `Sampler`, optional (default = `None`)
+            A sampler that can be used to select subsequent tokens.
+            If not given, search defaults to selecting the top `per_node_beam_size` tokens at each
+            step
+
+        # Returns
+
+        `BeamSearch` object with the defined parameters
+        """
+        return cls(
+            end_index=end_index,
+            max_steps=max_steps,
+            beam_size=beam_size,
+            per_node_beam_size=per_node_beam_size,
+            sampler=sampler,
+        )
+
+    @classmethod
+    def top_k_sampling(
+        cls,
+        end_index: int,
+        max_steps: int = 50,
+        beam_size: int = 10,
+        k: int = 1,
+        temperature: float = 1.0,
+    ) -> "BeamSearch":
+        """
+        Given an index of the end token in target vocabulary, return a BeamSearch object
+        that can be used to find `beam_size` candidate sequences, found by sampling from
+        the top `k` tokens to choose from at each step.
+
+        # Parameters
+
+        end_index : `int`
+            The index of the "stop" or "end" token in the target vocabulary.
+        max_steps : `int`, optional (default = `50`)
+            The maximum number of decoding steps to take, i.e. the maximum length
+            of the predicted sequences.
+        beam_size : `int`, optional (default = `10`)
+            The width of the beam used.
+        k : `int`, optional (default = `1`)
+            The number of top next tokens to sample from.
+        temperature : `float`, optional (default = `1.0`)
+            The 'temperature' of the probability distribution to be sampled
+            from, lowering below 1.0 creates a sharper distribution.
+
+        # Returns
+
+        `BeamSearch` object with the defined `end_index`, `max_steps`, and
+        `beam_size`, with subsequent tokens sampled using a `TopKSampler` with
+        the specified `k` and `temperature`
+        """
+        # Make sure `k` is a valid threshold
+        if type(k) is not int or k < 1:
+            raise ConfigurationError(
+                f'{"value of selection threshold `k` invalid."}'
+                f'{"`k` must be a positive `int`."}'
+            )
```

Comment:
What happens if `k` is greater than `beam_size` or `per_node_beam_size`? Will that break anything?

Reply:
Nope! Ideally k would be greater than beam size / per node beam size because you're sampling beam_size elements from the top k, and if k < beam size you're guaranteed repeats